### PR TITLE
Issue/3003 post enhancements

### LINKED ISSUE #3003: Post-related enhancements and fixes
* Consolidate the redundant `localDraft` and `uploaded` columns in the posts table
* Stop using `isUploaded` column in posts table, instead rely on the upload service to determine whether a post is uploading
* PostUploadService should request local draft from server after successful publish
* Use a service to request posts instead of requesting them inside PostsListFragment
* Resolve bug which causes dup rows after publishing a local draft (one for when it was a local draft, another for when it was published)

### PR DESCRIPTION
**Pull request by nbradbury**
Fixes #3003 - Introduces a few changes to the way posts are handled:

* The `uploaded` column in the posts table is no longer used (it's redundant since `localDraft` already tells us whether the post has been uploaded).

* The `isUploading` column in the posts table is no longer used in favor of asking `PostUpdateService` whether a post is uploading.

* `PostUploadService` now downloads local drafts after successful publish, to ensure the local copy matches the server.

* Adds `PostUpdateService` to update posts, replacing the current method of updating them from within `PostsListFragment`. Note that while this service is new, the code to request posts is the same as we've been using for a very long time.

* Fixes bug that failed to update the remote postid after a local draft was uploaded (prior to this fix there would be two rows in the db for the post - one for when it was a local draft and another for when it was published)

@roundhill If you're available, it'd be great to have you give this a look given your past experience with this code.

### CODE REVIEW

**Comment by nbradbury** [comment]
@roundhill If you're available, it'd be great to have you give this a look given your past experience with this code.

**Comment by roundhill** [comment]
Awesome, taking a look!

File: WordPress/src/main/java/org/wordpress/android/WordPressDB.java [hunk 1]
```diff
@@ -165,7 +164,7 @@
     //add boolean to posts to check uploaded posts that have local changes
     private static final String ADD_LOCAL_POST_CHANGES = "alter table posts add isLocalChange boolean default 0";
 
-    // Add boolean to POSTS to track posts currently being uploaded
+    // Add boolean to POSTS to track posts currently being uploaded - no longer used as of v4.5
     private static final String ADD_IS_UPLOADING = "alter table posts add isUploading boolean default 0";
```
Comment (roundhill) [inline WordPress/src/main/java/org/wordpress/android/WordPressDB.java]: Should we just delete this constant and put the note about deprecation in the version switch below where the column is added?

Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/WordPressDB.java]: Makes sense - added in c26474a. We devs should keep in mind, though, that switching from this branch to an earlier one may cause a crash due to the missing column.

File: WordPress/src/main/java/org/xmlrpc/android/ApiHelper.java [hunk 2]
```diff
@@ -1190,4 +1014,54 @@ else if (attrName.equals("href"))
...
+
+        try {
+            Object result = client.call(isPage ? "wp.getPage" : "metaWeblog.getPost", apiParams);
+
+            if (result != null && result instanceof Map) {
+                Map postMap = (HashMap) result;
+                List<Map<?, ?>> postsList = new ArrayList<>();
+                postsList.add(postMap);
+
+                WordPress.wpDB.savePosts(postsList, localBlogId, isPage, true);
```
Comment (roundhill) [inline WordPress/src/main/java/org/xmlrpc/android/ApiHelper.java]: Something's not quite working with the db `update` call in `savePosts`. From a quick debug I didn't see any problems with the WHERE clause, but it ends up with two rows for the Post for some reason:

screen shot 2015-07-24 at 4 42 56 pm

Comment (nbradbury) [inline WordPress/src/main/java/org/xmlrpc/android/ApiHelper.java]: Did that post already exist prior to trying this PR? If not, was it a brand new post or was it a local draft first?

Comment (roundhill) [inline WordPress/src/main/java/org/xmlrpc/android/ApiHelper.java]: > Did that post already exist prior to trying this PR? If not, was it a brand new post or was it a local draft first?

It didn't exist prior to the PR. Brand new post, published from the editor.

Comment (nbradbury) [inline WordPress/src/main/java/org/xmlrpc/android/ApiHelper.java]: I believe I found the problem. I was requesting the newly published post from the server before the existing post was updated in the local db (with the new remote id), resulting in dups due to the different IDs. And after looking at it, I decided we should request the uploaded post even if it wasn't a local draft to make sure the local copy matches the server.

Both of these are resolved in dd046e5.

**Comment by roundhill** [comment]
:shipit:

**Pull request merged**